In [34]:
import pandas as pd
import numpy as np

## Reading files

In [35]:
wh2018 = pd.read_csv('2018_v2.csv')
wh2019 = pd.read_csv('2019_v2.csv')

Adding extra column to identify year of population

In [36]:
wh2018['year'] = 2018
wh2019['year'] = 2019

Union of three files

In [37]:
wh = pd.concat([wh2018,wh2019])

In [38]:
wh.columns

Index(['Overall rank', 'Country or region', 'Score', 'GDP per capita',
       'Social support', 'Healthy life expectancy',
       'Freedom to make life choices', 'Generosity',
       'Perceptions of corruption', 'year'],
      dtype='object')

## Ejercicios

1. Partiendo de Wh, realice un analisis de valores nulos. que porcentaje de valores nulos se encuentran en cada columna por anio. presente el analisis en un dataframe

Ejemplo:

In [39]:
pd.DataFrame({
    "year": [2017,2017]
    , "column" : ["Happiness.Rank","Whisker.low"]
    , "percentage.of.nans" : [0.2,0.2]
})

year          column  percentage.of.nans
0  2017  Happiness.Rank                 0.2
1  2017     Whisker.low                 0.2

### resolucion ejercicio uno

In [40]:
dataFrame = pd.DataFrame(wh)
datos_nulos = dataFrame.isna().sum()
datos_nulos / 100

Overall rank                    0.00
Country or region               0.00
Score                           0.19
GDP per capita                  0.12
Social support                  0.16
Healthy life expectancy         0.26
Freedom to make life choices    0.23
Generosity                      0.21
Perceptions of corruption       0.13
year                            0.00
dtype: float64

se le asigno la variable dataframe = pandas (pddataframe) cree la variable datos nulos la iguale a dataframe utilice el metodo isna  con sum para hacer el recorrido y la suma en la tabla  luego dividi dicha suma entre 100 para obtener el porcentaje 

2. partiendo de Wh, por año muestre un resumen de todas las columnas.

nota: aunque solo tenemos tres años, que pasaria si en nuestro dataframe tenemos 20 años? resuelva su ejercicio lo mas dinamico que se pueda

### resolucion ejercicio dos

In [41]:
dataFrame = pd.DataFrame(wh)
allcolums = dataFrame.groupby(["year"], as_index=False).agg({"Overall rank": ["count"] ,"Score": ["count"],"Country or region": ["count"],  "GDP per capita": ["count"],
       "Social support": ["count"] ,  "Healthy life expectancy": ["count"],
       "Freedom to make life choices": ["count"], "Generosity": ["count"],
       "Perceptions of corruption": ["count"] })


allcolums



year Overall rank Score Country or region GDP per capita Social support  \
               count count             count          count          count   
0  2018          156   151               156            147            141   
1  2019          156   142               156            153            155   

  Healthy life expectancy Freedom to make life choices Generosity  \
                    count                        count      count   
0                     144                          144        139   
1                     142                          145        152   

  Perceptions of corruption  
                      count  
0                       149  
1                       150

se hizo un conteo de los elementos en la variable allcolums del dataframe despues los agrupe con un groupby y los mande a imprimir 

3. partiendo de Wh, remplace los valores NaN de las columnas por su mean, el mean debe de ser calculado por año para ser mas exacto

### resolucion ejercicio tres

4. Partiendo de Wh limpio, para cada metrica por pais, muestra la diferencia del ultimo valor encontrado contra el primero. 

Ejemplo:

In [42]:
wh.loc[ wh['Country or region'] == 'Switzerland' , : ]

Overall rank Country or region  Score  GDP per capita  Social support  \
4             5       Switzerland  7.487           1.420           1.549   
5             6       Switzerland  7.480           1.452           1.526   

   Healthy life expectancy  Freedom to make life choices  Generosity  \
4                    0.927                         0.660       0.256   
5                    1.052                         0.572       0.263   

   Perceptions of corruption  year  
4                      0.357  2018  
5                      0.343  2019

Como se puede observar para Switzerland en Score la diferencia seria 7.480 - 7.487

Claramente en su solucion la columna año se debera de eliminar

### resolucion ejercicio cuatro